oemof_b3/tools/data_processing.py

```diff
@@ -1,7 +1,761 @@
+                if value == df[key][index]:
+                    df_filtered_by_value = df_filtered_by_value.append(
+                        row, ignore_index=True
+                    )
+
+    return df_filtered_by_value
+
+
+def df_agg(df, key):
+    """
+    This function aggregates columns of a DataFrame
+
+    Parameters
+    ----------
+    df : pd.DataFrame
+        DataFrame
+    key : string
+        The column's name aggregated by
+
+    Returns
+    -------
+    df_agg : pd.DataFrame
+        DataFrame with aggregated columns
+    """
+
+    # Get header of scalars
+    scalars_header = get_optional_required_header("scalars")
+    header_scalars = scalars_header[0]
+    optional_header_scalars = scalars_header[1]
+    required_header_scalars = scalars_header[2]
+
+    # Get header of time series
+    timeseries_header = get_optional_required_header("timeseries")
+    header_timeseries = timeseries_header[0]
+    optional_header_timeseries = timeseries_header[1]
+    required_header_timeseries = timeseries_header[2]
+
+    # Save header of DataFrame to variable
+    df_header = list(df.columns)
+
+    df_header_required = df_header.copy()
+    for item in df_header:
+        if item in optional_header_scalars:
+            df_header_required.remove(item)
+        elif item in optional_header_timeseries:
+            df_header_required.remove(item)
+
+    scalars = False
+    time_series = False
+
+    if df_header_required == required_header_scalars:
+
+        # Aggregation options for scalars
+        scalars = True
+        agg_options = ["region", "carrier", "tech"]
+
+    elif df_header_required == required_header_timeseries:
+        time_series = True
+        # Aggregation options for time series
+        agg_options = ["region"]
+
+    else:
+        newline = "\n"
+        raise KeyError(
+            f"The data you passed is neither a stacked time series nor does it contain scalars. "
+            f"{newline}"
+            f"Please make sure your data contains the following columns {newline}"
+            f"time series: {header_timeseries}{newline}"
+            f"scalars: {header_scalars}{newline}"
+        )
+
+    # Ensure key is a valid aggregation option
+    if key not in agg_options:
+        raise KeyError(
+            f"{key} is not a option for a aggregation."
+            f"Please choose of one of these aggregation options: {agg_options}"
+        )
+
+    # Check if key is in header
+    if key not in df_header:
+        raise KeyError(
+            f"Your data is missing the column {key}."
+            f"Please provide a complete data set with the required column"
+        )
+
+    # Empty DataFrame, which will contain aggregated items
+    df_agg_by_key = pd.DataFrame(columns=list(df.columns))
+
+    # Get list of all items existing in the column passed with key
+    key_list = list(df[key])
+    key_list = list(dict.fromkeys(key_list))
+
+    # Begin aggregation of scalars
+    if scalars:
```

Comment:
This is a hard-to-review part starting here: Deeply nested loops, lots of values are set. I would need more time to understand this. Hard to maintain as well. We should try to drastically simplify this. Let's talk about it.